code/modules/mob/living/carbon/xenomorph/castes/praetorian/abilities_praetorian.dm

```diff
@@ -467,3 +466,454 @@ GLOBAL_LIST_INIT(acid_spray_hit, typecacheof(list(/obj/structure/barricade, /obj
+	initial_turf = get_step(xeno_owner, xeno_owner.dir)
+	turf_line = get_turf_line(initial_turf, A, 7)
+	LAZYINITLIST(telegraphed_atoms)
+	for(var/turf/turf_from_line AS in turf_line)
+		telegraphed_atoms += new /obj/effect/xeno/abduct_warning(turf_from_line)
+	ADD_TRAIT(xeno_owner, TRAIT_IMMOBILE, XENO_TRAIT)
+	ability_timer = addtimer(CALLBACK(src, PROC_REF(pull_them_in)), 1.2 SECONDS, TIMER_STOPPABLE|TIMER_UNIQUE)
+	RegisterSignal(xeno_owner, COMSIG_MOVABLE_MOVED, PROC_REF(failed_pull))
+	RegisterSignal(xeno_owner, COMSIG_LIVING_STATUS_STAGGER, PROC_REF(failed_pull))
+
+/// Get a filtered line of turfs from a turf to a target.
+/datum/action/ability/activable/xeno/abduct/proc/get_turf_line(turf/starting_turf, atom/movable/target, distance)
+	var/list/turf_line_unfiltered = getline(starting_turf, target)
+	var/list/turf_line_filtered = list()
+	for(var/turf/unfiltered_turf AS in turf_line_unfiltered)
+		if(turf_line_filtered.len > distance)
+			break
+		if(unfiltered_turf.density)
+			break
+		var/blocked = FALSE
+		for(var/obj/object_on_turf in unfiltered_turf)
+			if(isbarricade(object_on_turf))
+				var/obj/structure/barricade/barricade_on_turf = object_on_turf
+				if(!barricade_on_turf.climbable)
+					blocked = TRUE
+					break
+				continue
+			if(object_on_turf.density && !(object_on_turf.allow_pass_flags & PASS_THROW))
+				blocked = TRUE
+				break
+		if(blocked)
+			break
+		turf_line_filtered += unfiltered_turf
+
+	return turf_line_filtered
+
+/// Ends the ability by throwing all humans in the affected turfs to the initial turf.
+/datum/action/ability/activable/xeno/abduct/proc/pull_them_in()
+	SIGNAL_HANDLER
+	var/list/mob/living/carbon/human/human_mobs = list()
+	for(var/turf/turf_from_line AS in turf_line)
+		for(var/atom/movable/target AS in turf_from_line.contents)
+			if(!ishuman(target))
+				continue
+			var/mob/living/carbon/human/human_mob = target
+			if(human_mob.stat == DEAD)
+				continue
+			human_mobs += target
+
+	for(var/mob/living/carbon/human/human_mob in human_mobs)
+		human_mob.throw_at(owner, 6, 2, initial_turf, FALSE)
+		human_mob.Paralyze(0.1 SECONDS)
+		human_mob.add_slowdown(0.3 * human_mobs.len)
+		human_mob.adjust_stagger(0.5 SECONDS * human_mobs.len)
+		human_mob.apply_effect(human_mobs.len >= 3 ? 1 SECONDS : 0.1 SECONDS, WEAKEN)
+	if(human_mobs.len)
+		xeno_owner.add_slowdown(0.4 * human_mobs.len) // Don't bite off more than what you can chew.
+		playsound(human_mobs[human_mobs.len], 'sound/voice/alien/pounce.ogg', 25, TRUE)
+	succeed_activate()
+	add_cooldown()
+	cleanup_variables()
+
+/// Ends the ability by punishing the owner.
+/datum/action/ability/activable/xeno/abduct/proc/failed_pull()
+	SIGNAL_HANDLER
+	xeno_owner.Knockdown(1 SECONDS)
+	xeno_owner.add_slowdown(0.9)
+	succeed_activate()
+	add_cooldown()
+	cleanup_variables()
+
+/// Cleans up any variables, signals, and undoes any traits that the ability gave.
+/datum/action/ability/activable/xeno/abduct/proc/cleanup_variables()
```

Comment:
You should call this on destroy, since it has various refs that need cleaning up if the xeno dies mid cast.